proxy/server/server.go

```diff
@@ -357,3 +377,144 @@ func newScopeRouter(scope string, router http.Handler) (http.Handler, error) {
+				}
+			} else if block.Type == "EC PRIVATE KEY" {
+				ecPrivateKey, err := x509.ParseECPrivateKey(block.Bytes)
+				if err == nil {
+					autoCertMgr.Client.Key = ecPrivateKey
+				} else {
+					log.Errorf("Error parsing Autocert Key block of type %s, for Host %s, as an EC Private Key: %v", block.Type, hostCfg.Name, err)
+				}
+			} else {
+				return nil, fmt.Errorf("AutoCert Private Key for Host %s is of unrecognized type: %s. Supported types"+
+					"are RSA PRIVATE KEY and EC PRIVATE KEY.", hostCfg.Name, block.Type)
+			}
+		}
+	}
+
+	// Wrap the GetCert for this host, so we can generate and staple
+	// an optional OCSP response when requested.
+	stapledGetCert := func(info *tls.ClientHelloInfo) (*tls.Certificate, error) {
+		keyPair, err := autoCertMgr.GetCertificate(info)
+
+		if hostCfg.Settings.OCSP.Enabled {
+			log.Infof("OCSP is enabled for %v, resolvers: %v", hostCfg, hostCfg.Settings.OCSP.Responders)
+			r, err := s.StapleHost(&hostCfg, stapler.WithGetCertFunc(hostCfg.Name, stapler.GetCertificateFunc(autoCertMgr.GetCertificate)))
+			if err != nil {
+				log.Warningf("Failed to staple %v, error %v", hostCfg, err)
+			} else if r.Response.Status == ocsp.Good || r.Response.Status == ocsp.Revoked {
+				keyPair.OCSPStaple = r.Staple
+			} else {
+				log.Warningf("Got undefined status from OCSP responder: %v", r.Response.Status)
+			}
+		}
+		return keyPair, err
+	}
+
+	return stapledGetCert, nil
+}
+
+// Generate an aggregate GetCertificate function over GetCertificate functions for
+// each host that we generated using the certFuncForHost function above.
+// This allows all of those functions to masquerade as one uber function that can
+// get a certificate (through autocert) for any host.
+func getCertFuncAggregate(getCertFuncs map[string]getCertificateFunc) getCertificateFunc {
+	return func(info *tls.ClientHelloInfo) (*tls.Certificate, error) {
+		if getCertificateFunc, ok := getCertFuncs[info.ServerName]; ok {
+			// We have a get certificate function for this host - allow AutoCertManager to
+			// provide this one, in case there's expiry/renewal to be done.
+			cert, err := getCertificateFunc(info)
+			if err != nil {
+				log.Errorf("Failed to generate Autocert for ServerName: %s. Error: %v.", info.ServerName, err)
+			}
+			return cert, err
+		} else {
+			return nil, nil
+		}
+	}
+}
+
+func tlsCertArray(pairs map[string]tls.Certificate, defaultHostName string) ([]tls.Certificate, error) {
+	arr := make([]tls.Certificate, 0, len(pairs))
+	if defaultHostName != "" {
+		keyPair, ok := pairs[defaultHostName]
+		if !ok {
+			return nil, errors.Errorf("default host '%s' certificate is not passed", defaultHostName)
+		}
+		arr = append(arr, keyPair)
+	}
+
+	for hostName, keyPair := range pairs {
+		if hostName != defaultHostName {
+			arr = append(arr, keyPair)
+		}
+	}
+	return arr, nil
+}
+
+func ocspStapleToCert(stapler stapler.Stapler, hostCfg engine.Host, keyPair *tls.Certificate) {
```

Comment:
Code with nested if and if else is hard to read. It is more Golang idiomatic to return as soon as you have a chance, How about:
```Go
func ocspStapleToCert(stapler stapler.Stapler, hostCfg engine.Host, keyPair *tls.Certificate) {
	if !hostCfg.Settings.OCSP.Enabled {
		return
	}
	log.Infof("OCSP is enabled for %v, resolvers: %v", hostCfg, hostCfg.Settings.OCSP.Responders)

	r, err := stapler.StapleHost(&hostCfg)
	if err != nil {
		log.Warningf("Failed to staple %v, error %v", hostCfg, err)
		return
	} 
	if r.Response.Status != ocsp.Good && r.Response.Status != ocsp.Revoked {
		log.Warningf("Got undefined status from OCSP responder: %v", r.Response.Status)
		return
	}
	keyPair.OCSPStaple = r.Staple
}
```

Reply:
Did that...